sql/analyzer/costed_index_scan.go

```diff
@@ -1286,153 +1286,165 @@ func (c *indexCoster) costFulltext(filter *iScanLeaf, s sql.Statistic, ordinal i
+	IndexScanOpInSet                         // =
+	IndexScanOpNotInSet                      // !=
+	IndexScanOpNotEq                         // !=
+	IndexScanOpGt                            // >
+	IndexScanOpGte                           // >=
+	IndexScanOpLt                            // <
+	IndexScanOpLte                           // <=
+	IndexScanOpAnd                           // &&
+	IndexScanOpOr                            // ||
+	IndexScanOpIsNull                        // IS NULL
+	IndexScanOpIsNotNull                     // IS NOT NULL
+	IndexScanOpSpatialEq                     // SpatialEq
+	IndexScanOpFulltextEq                    // FulltextEq
 )
 
-// swap returns the identity op for swapping a comparison's LHS and RHS
-func (o indexScanOp) swap() indexScanOp {
+// Swap returns the identity op for swapping a comparison's LHS and RHS
+func (o IndexScanOp) Swap() IndexScanOp {
 	switch o {
-	case indexScanOpGt:
-		return indexScanOpLt
-	case indexScanOpGte:
-		return indexScanOpLte
-	case indexScanOpLt:
-		return indexScanOpGt
-	case indexScanOpLte:
-		return indexScanOpGte
+	case IndexScanOpGt:
+		return IndexScanOpLt
+	case IndexScanOpGte:
+		return IndexScanOpLte
+	case IndexScanOpLt:
+		return IndexScanOpGt
+	case IndexScanOpLte:
+		return IndexScanOpGte
 	default:
 		return o
 	}
 }
 
 func newLeaf(ctx *sql.Context, id indexScanId, e sql.Expression, underlying string) (*iScanLeaf, bool) {
-	var op indexScanOp
+	op, left, right, ok := IndexLeafChildren(e)
+	if !ok {
+		return nil, false
+	}
+	if op == IndexScanOpFulltextEq {
+		e := e.(*expression.MatchAgainst)
+		return &iScanLeaf{id: id, op: op, gf: e.Columns[0].(*expression.GetField), underlying: underlying, fulltextIndex: e.GetIndex().ID()}, true
+	}
+	if _, ok := left.(*expression.GetField); !ok {
+		left, right = right, left
+		op = op.Swap()
+	}
+
+	gf, ok := left.(*expression.GetField)
+	if !ok {
+		return nil, false
+	}
+
+	if op == IndexScanOpIsNull || op == IndexScanOpIsNotNull {
+		return &iScanLeaf{id: id, gf: gf, op: op, underlying: underlying}, true
+	}
+
+	if !isEvaluable(right) {
+		return nil, false
+	}
+
+	if op == IndexScanOpInSet || op == IndexScanOpNotInSet {
+		tup := right.(expression.Tuple)
+		var litSet []interface{}
+		for _, lit := range tup {
+			value, err := lit.Eval(ctx, nil)
+			if err != nil {
+				return nil, false
+			}
+			litSet = append(litSet, value)
+		}
+		return &iScanLeaf{id: id, gf: gf, op: op, setValues: litSet, underlying: underlying}, true
+	}
+
+	value, err := right.Eval(ctx, nil)
+	if err != nil {
+		return nil, false
+	}
+
+	return &iScanLeaf{id: id, gf: gf, op: op, litValue: value, underlying: underlying}, true
+}
+
+// IndexLeafChildren handles the struct types that may be found on a leaf node while creating indexes. Integrators may
+// change this function to allow for different struct types.
+var IndexLeafChildren = func(e sql.Expression) (IndexScanOp, sql.Expression, sql.Expression, bool) {
```

Comment:
zach tells me you will not be dissuaded from this pattern, but it's also not my favorite

Reply:
Zach mentioned that he was going to find a better solution that isn't threading interfaces everywhere, but hasn't mentioned anything yet.